internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -477,15 +478,22 @@ private[inc] abstract class IncrementalCommon(
     if (secondClassInvalidation.nonEmpty)
       log.debug(s"Invalidated due to generated class file collision: ${secondClassInvalidation}")
 
-    val newInvalidations = (firstClassInvalidation -- recompiledClasses) ++ secondClassInvalidation
+    // Invalidate macros that transitively depend on any of the recompiled classes
+    val thirdClassInvalidation = {
```

Comment:
LGTM in terms of transitive dependency correctness. There's probably still some edge cases left but this did cover the edge cases I can think of.

Preferably someone can do a performance review. If the performance is acceptable then I personally think this is good to go. If not, we might need to think of a way to optimize computation of `thirdClassInvalidation`, or fall back to initial approach of checking at `invalidateInitial` only.

Reply:
I don't think computing `thirdClassInvalidation` would impact performance, because `invalidateClassesInternally` already calls `transitiveDeps` - although there it operates on the classes in `changes`, which is a subset of `recompiledClasses` we're operating on here (because we care about classes that changed in behaviour, not just those that changed in API).  Let me know if I'm missing something.